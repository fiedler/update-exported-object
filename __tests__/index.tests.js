// @flow

const updateExportedObject = require("../src");

const file = `
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: [
      { "hello": "world" }
    ],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
`;

type Something = {
  anObject: {
    oldProperty: number
  },
  anArray: Array<{ [string]: string }>,
  isSomething: boolean,
  aFunc: typeof Symbol
};

test("create object", () => {
  updateExportedObject(file, (obj: Something) => {
    expect(obj.anObject.oldProperty).toBe(2);
    expect(obj["anArray"].length).toBe(1);
    expect(obj.isSomething).toBe(true);
    expect(obj.aFunc.toString()).toBe("Symbol(unknown)");
  });
});

test("invalid javascript", () => {
  expect(() => {
    updateExportedObject("blah blah", () => {});
  }).toThrowError(/Unexpected identifier/);
});

test("no module.exports", () => {
  expect(() => {
    updateExportedObject("var x = 0;", () => {});
  }).toThrowError(/module.exports is not of type 'object'/);
});

test("module.exports is wrong type", () => {
  expect(() => {
    updateExportedObject("module.exports = 1", () => {});
  }).toThrowError(/module.exports is not of type 'object'/);
});

test("update property in object", () => {
  expect(
    updateExportedObject(file, obj => {
      obj.anObject.oldProperty = 1;
      obj.isSomething = false;
      obj.aFunc = { change: "type" };
    })
  ).toMatchSnapshot();
});

test("insert property in object", () => {
  expect(
    updateExportedObject(file, obj => {
      obj.anObject.newProperty = 1;
      obj.isNull = null;
    })
  ).toMatchSnapshot();
});

test("insert property with wrong type", () => {
  expect(() => {
    updateExportedObject(file, obj => {
      obj.newFunc = () => {};
    });
  }).toThrowError(/value type '\w+' not supported/);
});

test("delete property from object", () => {
  expect(
    updateExportedObject(file, obj => {
      delete obj.anObject["oldProperty"];
    })
  ).toMatchSnapshot();
});

test("get element from array", () => {
  updateExportedObject(file, obj => {
    expect(obj.anArray[0].hello).toBe("world");
  });
});

test("set element from array", () => {
  updateExportedObject(file, obj => {
    expect(
      updateExportedObject(file, obj => {
        obj.anArray[0] = { yes: "please" };
      })
    ).toMatchSnapshot();
  });
});

test("splice from array", () => {
  updateExportedObject(file, obj => {
    expect(
      updateExportedObject(file, obj => {
        obj.anArray.splice(0, 1);
      })
    ).toMatchSnapshot();
  });
});

test("push to array", () => {
  updateExportedObject(file, obj => {
    expect(
      updateExportedObject(file, obj => {
        obj.anArray.push("hello");
      })
    ).toMatchSnapshot();
  });
});

test("change type", () => {
  updateExportedObject(file, obj => {
    expect(
      updateExportedObject(file, obj => {
        obj.anArray = 1;
      })
    ).toMatchSnapshot();
  });
});

test("change type in array", () => {
  updateExportedObject(file, obj => {
    expect(
      updateExportedObject(file, obj => {
        obj.anArray[0] = "hej";
      })
    ).toMatchSnapshot();
  });
});
