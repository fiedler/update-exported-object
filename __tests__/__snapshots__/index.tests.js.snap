// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change type 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: 1,
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`change type in array 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: [
      \\"hej\\"
    ],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`delete property from object 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {},
    anArray: [
      { \\"hello\\": \\"world\\" }
    ],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`insert property in object 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      // there is a comment here
      oldProperty: 2,

      newProperty: 1
    },

    anArray: [
      { \\"hello\\": \\"world\\" }
    ],

    isSomething: true,
    aFunc: callback,
    isNull: null
  }

  function callback() {

  }
"
`;

exports[`push to array 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: [{ \\"hello\\": \\"world\\" }, \\"hello\\"],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`set element from array 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: [
      { yes: \\"please\\" }
    ],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`splice from array 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 2  // there is a comment here
    },
    anArray: [],
    isSomething: true,
    aFunc: callback
  }

  function callback() {

  }
"
`;

exports[`update property in object 1`] = `
"
  var thereIsSomethingBefore = true;

  module.exports = {
    anObject: {
      oldProperty: 1  // there is a comment here
    },
    anArray: [
      { \\"hello\\": \\"world\\" }
    ],
    isSomething: false,
    aFunc: {
      change: \\"type\\"
    }
  }

  function callback() {

  }
"
`;
